### ISSUE TITLE
Parallelize *.pom fetching?

### ISSUE THREAD

**Issue by Alan47** [ben-manes/gradle-versions-plugin#262]
This plugin is very useful. We use it on a relatively large project where we have tons of dependencies. This unfortunately also means that the `dependencyUpdates` task runs for a good 30 minutes (!). Most of the time, the process is waiting for a POM file download to finish.

I think the process would benefit a lot from parallel POM fetching instead of downloading them one after the other. Do you think that this might be possible?

**Comment by ben-manes** [comment]
I agree parallel metadata resolution would be wonderful, but alas this is a missing Gradle feature. You should open an issue on their repository.

This plugin script's Gradle's resolution handling, which is both much simpler and more accurate representation of your build (with regards to your resolution strategies, repository proxy settings, etc). Unfortunately that means we are limited to what Gradle provides and beholden to their development. If we wrote our own resolution logic it could be faster, but more work with potential incompatibilities. It seems better for everyone to see Gradle improve so that all use-cases benefit.

We tried to add some parallelization by resolving configurations concurrently. Unfortunately that can deadlock and we may have to remove it (#255), per the Gradle team's advise against our approach.

In the past I have asked Gradle to resolve dependencies in parallel. They finally did so, but only for the jars and not the metadata. At the time, the developers said it was possible but lower priority so they did not complete the work.

You may want to consider strategies using a CI task to run the plugin and publish the report. This way the long execution doesn't impact anyone's work by being a hidden, background process.

**Comment by Alan47** [comment]
Thank you for the explanation. I wasn't aware that this is a limitation of Gradle itself. It is pretty sad that a state-of-the-art tool can't do this kind of thing in parallel. I can imagine though that the parallelization process *in general* is not an easy problem: what POM you need next may depend on the contents of the current one. However, for the special case of checking for outdated dependencies, it is much simpler because the set of all POMs is already known. Sending a request for the latest POM version for each of them should be easier. But I can definitly see why you want to stick with the default API. Quite the dilemma here...

**Comment by ben-manes** [comment]
You may want to refer to https://github.com/gradle/gradle/issues/1309 (and @bigdaz who did the work). In particular the comment,
> The release notes aren't entirely correct. Gradle only downloads metadata in parallel during resolution of the dependency graph. It still does version selection sequentially, and this will include the network requests to determine the versions that are available to match a particular dynamic version. So in your case I would expect everything to still happen sequentially. This is something we will improve later.

For most builds it is a one-time cost with only a few dynamic versions to resolve, and the internal cache then avoids the penalty thereafter. Of course this plugin breaks that assumption entirely, but is a rare case. Its then hard to argue for the improvements given their other tasks. Ideally they would build this plugin's feature into the distribution, which might then incentivize that performance improvement.

Until then, I think using a CI task is likely the best strategy to keep it from being user-facing.

**Comment by Alan47** [comment]
>  Ideally they would build this plugin's feature into the distribution, which might then incentivize that performance improvement.

This would definitly be the best solution. Dependency management is one of Gradle's core capabilities, and I think that looking for updates is part of this capability. Thanks for your work on this plugin, I can't even imagine doing this manually every time.

**Comment by MartinX3** [comment]
Does gradle 5.0 support this now?

**Comment by MartinX3** [comment]
Does gradle 5.4 support this now?

**Comment by ben-manes** [comment]
Nope
